### PR TITLE
[SPARK-14201][SQL] handle empty file for new data source strategy

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

It's possible that users will write empty files to data dir. For example, write data to JSON data source with one line string, and default parallelism is 2, we will write an empty file and a file with all data. The new data source should be able to handle it.
## How was this patch tested?

new tests in `FileSourceStrategySuite`

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @liancheng @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 1]
```diff
@@ -549,15 +549,18 @@ class HDFSFileCatalog(
 
   refresh()
 
+  private def isInvalidFile(f: FileStatus): Boolean =
+    f.getPath.getName.startsWith("_") || f.getLen == 0
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Where is the place that does not handle empty file properly? Does that also handle file with no records (a csv file has only header line)

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Yeah, its not clear to me that we want to skip 0 sized files.  They really shouldn't be there for valid datasets.  What is the motivation?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I would say that we should skip writing out empty files.

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Even we can guarantee that the writer in Spark will not create empty file, but other database (Hive) may do that. We should still handle that in Spark.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: The new FileSourceStratyge expects non-empty files [here](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala#L141)

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Why not filter out the empty files there?

Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Actually there is an existing [CSV test case][1] that cover empty CSV input files. I have to [remove the empty file assertion][2] in `FileSourceStrategy` for this test case when implementing `buildReader()` for CSV.

[1]: https://github.com/apache/spark/blob/e5a1b301fbe191f1a9627a1083d960c98f543d13/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala#L248-L256
[2]: https://github.com/apache/spark/pull/12002#discussion_r57594679

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: I wanna filter out them earlier, so I choose here. Is empty file valid in some cases?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: It's correct to be here, just confusing.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Yeah, that assert is my fault (from when I though my bin packing algorithm had a bug).  We should support reading empty files (if thats valid for the format) and we should avoid writing them when possible.

**Comment by cloud-fan** [comment]
The fix is so simple, and we already have tests for empty files, so I'm going to close this PR and the issue is fixed in https://github.com/apache/spark/pull/11960

**Pull request closed without merging**